src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -919,19 +920,29 @@ query_path(const char *szInput, int *is_target_path, dfs_obj_t **parent, char *i
 				D_GOTO(out_err, rc = ENAMETOOLONG);
 			}
 			strncpy(item_name, full_path_parse + pos + 1, len_item_name + 1);
+
 			/* the item under root directory */
 			if (pos == (*dfs_mt)->len_fs_root) {
+				/* TODO: Do we need to call lookup_insert_dir() in this case, as
+				 * there's no ref counting on the hash table it's probably
+				 * sufficient to keep a pointer to the root dfs object and use
+				 * that here.
+				 */
 				*parent       = NULL;
 				(*parent_dir)[0] = '/';
-				(*parent_dir)[1] = '\0';
+				path_len         = 1;
```

Comment:
Do you intend for the DFS root to be in the hash table or not?  I think most DFS options can take NULL for a parent and then assume the root but it would certainly be more consistent if we always passed oh - and that's what the existing code does although the fact you set *parent to NULL here makes me wonder if that was intentional or not.

Assuming we do want / in the hash table then I suspect we could just avoid this conditional and just do the strncpy in either case which would simplify the code.

In terms of the bigger picture I the query_path function is passing out an allocated parent_dir here but actually it could return a pointer to the hash table entry, then the string would not need to be allocated or copied (or freed by the callers) but that's a bigger change and I was trying to stick to bug-fixes here.

Finally, I think overall this function would be better if it set `*parent = NULL` on the way in and then only overwrite it if there is a valid value for it.  Right now it's set to null in most (or all cases?) and sometimes set to NULL and then overwritten.  If you did this then you could remove the `*ojb = NULL` in lookup_insert_dir() as well and the code would be safer overall.

Reply:
> Regarding "return a pointer to the hash table entry", are the location of each entry fixed or they could be re-positioned or even evicted?

They will be fixed and without reference counting then they won't be evicted either.  I think the best here would be for lookup_insert_dir and query_path to return a hash table entry with reference held and have the caller call decref() when done but that's a bigger change.

Reply:
> I do not remember exactly now. Yes. I agree that we can return after "path_len = 1;" and simplify else {}.

I think it would be better for this function to consistently return a `dfs_obt_t *` for now even if it may not be strictly needed.